mp4e/mp4.go

```diff
@@ -0,0 +1,308 @@
+			//}
+
+			if j == 0 {
+				segInfo.SeqStart = seq
+			} else if j == len(seg.Fragments)-1 {
+				segInfo.SeqEnd = seq
+			}
+
+			if seq != seqPrev+1 && j > 0 {
+				info.AddError(fmt.Sprintf("Sequence number gap %d - %d", seqPrev, seq), i, j, -1)
+			}
+			seqPrev = seq
+
+			var samples []mp4.FullSample
+			samples, err = frag.GetFullSamples(nil)
+			if err != nil {
+				info.AddError(fmt.Sprintf("failed to read samples: %v", err), i, j, -1)
+				continue
+			}
+			sampleCount := uint64(len(samples))
+			segInfo.SampleCount += sampleCount
+			for k, sample := range samples {
+				dur := uint64(sample.Dur)
+				if dur < info.SampleDurationMin || info.SampleDurationMin == 0 {
+					info.SampleDurationMin = dur
+				}
+				if dur > info.SampleDurationMax {
+					info.SampleDurationMax = dur
+				}
+				dts := sample.DecodeTime
+				pts := sample.PresentationTime()
+				segInfo.Samples = append(segInfo.Samples, &SampleInfo{dur, j, pts})
+				if j == 0 && k == 0 {
+					segInfo.DtsStart = dts
+					segInfo.PtsStart = pts
+					if err = ensureIDRFrame(sample); err != nil {
+						info.AddError(err.Error(), i, j, int64(pts))
+					}
+				} else if j == len(seg.Fragments)-1 && k == len(samples)-1 {
+					segInfo.DtsEnd = dts + dur
+					segInfo.PtsEnd = pts + dur
+				}
+				if dtsPrev != 0 && absDiff(dts-dtsPrev, dur) > 1 {
+					info.AddError(fmt.Sprintf("DTS gap %d - %d, sample duration %d", dtsPrev, dts, dur), i, j, int64(pts))
+				}
+				dtsPrev = dts
+			}
+		}
+
+		// PTS may not be in order
+		sort.Slice(segInfo.Samples, func(a, b int) bool {
+			return segInfo.Samples[a].Pts < segInfo.Samples[b].Pts
+		})
+		for _, s := range segInfo.Samples {
+			if ptsPrev != 0 && absDiff(s.Pts-ptsPrev, s.Duration) > 1 {
+				info.AddError(fmt.Sprintf("PTS gap %d - %d, sample duration %d", ptsPrev, s.Pts, s.Duration), i, s.Fragment, 0)
+			}
+			ptsPrev = s.Pts
+		}
+
+		info.SampleCount += segInfo.SampleCount
+		if isSampleCountSequenceBad(sampleCountPrevPrev, sampleCountPrev, segInfo.SampleCount) &&
+			i < len(file.Segments)-1 { // ignore last segment
+			info.AddError(fmt.Sprintf("Sample count mismatch %d - %d", sampleCountPrev, segInfo.SampleCount), i, -1, 0)
+		}
+		sampleCountPrevPrev = sampleCountPrev
+		sampleCountPrev = segInfo.SampleCount
+		if segInfo.SampleCount < info.SampleCountMin || info.SampleCountMin == 0 {
+			info.SampleCountMin = segInfo.SampleCount
+		}
+		if segInfo.SampleCount > info.SampleCountMax {
+			info.SampleCountMax = segInfo.SampleCount
+		}
+		info.Segments = append(info.Segments, segInfo)
+	}
+
+	return file, info, nil
+}
+
+func absDiff(x, y uint64) uint64 {
+	if x < y {
+		return y - x
+	}
+	return x - y
+}
+
+// ensureIDRFrame returns nil if the sample contains an IDR frame, or is not
+// a video sample. Assumes AVC encoding
+func ensureIDRFrame(sample mp4.FullSample) (err error) {
+	nalus, err := avc.GetNalusFromSample(sample.Data)
+	if err != nil {
+		if strings.Index(err.Error(), "video") != -1 {
+			err = nil
```

Comment:
By looking at the code, this means that mp4ff thinks that the input is not video? So we don't care about IDR frames as long as its not video. Do you think that this could possibly result in some form of bad video being ignored if it is also broken in this way? Can we check that this sample _should_ be video?

Reply:
Yeah the error message is "NALU length fields are bad. Not video?". I'll check for "Not video?" instead for now.

I don't know how to easily figure out if a file is video or audio actually, if it doesn't contain a `moov` atom.